functions/go/ensure-name-segment/ensure_name_segment.go

```diff
@@ -0,0 +1,203 @@
+package main
+
+import (
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/kustomize/api/filters/prefixsuffix"
+	"sigs.k8s.io/kustomize/api/resid"
+	"sigs.k8s.io/kustomize/api/resmap"
+	"sigs.k8s.io/kustomize/api/resource"
+	"sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+	k8syaml "sigs.k8s.io/yaml"
+)
+
+const (
+	fnConfigGroup      = "fn.kpt.dev"
+	fnConfigVersion    = "v1alpha1"
+	fnConfigAPIVersion = fnConfigGroup + "/" + fnConfigVersion
+	fnConfigKind       = "EnsureNameSegment"
+)
+
+type EnsureNameSegment struct {
+	yaml.ResourceMeta `json:",inline" yaml:",inline"`
+	// Segment is the desired name segment.
+	Segment string `json:"segment" yaml:"segment"`
+	// ActionOnNotFound controls the desired action when the desired segment is not found in the name.
+	// If not specified, prepend will be the default.
+	ActionOnNotFound ActionMode `json:"actionOnNotFound,omitempty" yaml:"actionOnNotFound,omitempty"`
+
+	FieldSpecs []types.FieldSpec `json:"fieldSpecs,omitempty" yaml:"fieldSpecs,omitempty"`
+}
+
+type ActionMode string
+
+const (
+	Prepend ActionMode = "prepend"
+	Append  ActionMode = "append"
+)
+
+func (ens *EnsureNameSegment) Defaults() {
+	if ens.ActionOnNotFound == "" {
+		ens.ActionOnNotFound = Prepend
+	}
+}
+
+func (ens *EnsureNameSegment) Validate() error {
+	if len(ens.Segment) == 0 {
+		return fmt.Errorf("segment must not be empty")
+	}
+	return nil
+}
+
+func (ens *EnsureNameSegment) Transform(m resmap.ResMap) error {
+	for _, r := range m.Resources() {
+		if shouldSkip(r.OrgId()) {
+			continue
+		}
+		id := r.OrgId()
+		// current default configuration contains
+		// only one entry: "metadata/name" with no GVK
+		for _, fs := range ens.FieldSpecs {
+			// TODO: this is redundant to filter (but needed for now)
+			if !id.IsSelected(&fs.Gvk) {
+				continue
+			}
+
+			// Idempotency check: if the segment is already part of the name, we
+			// don't need to do anything.
+			contain, err := resourceContainsSegment(r, ens.Segment, fs)
+			if err != nil {
+				return err
+			}
+			if contain {
+				continue
+			}
+
+			if isNameChange(&fs) {
+				// If we are changing "metadata/name", we tracks the original
+				// name and the prefix or suffix being added.
+				r.StorePreviousId()
```

Comment:
So its reusing kustomize's implementation. That gives us good confidence in the functionality.

Reply:
Yes, we try to reuse kustomize's implementation as much as possible.